### PR TITLE
[UBSan] Diagnose assumption violation

### PR DESCRIPTION
**Pull request by dtcxzyw**
This patch extends [D34590](https://reviews.llvm.org/D34590) to check assumption violations.

### CODE REVIEW

**Comment by dtcxzyw** [comment]
Ping.

**Comment by dtcxzyw** [comment]
Ping.

**Comment by fmayer** [comment]
LGTM, but would like @vitalybuka to also take a quick look.

File: clang/lib/CodeGen/CGBuiltin.cpp [hunk 1]
```diff
@@ -1996,16 +1996,21 @@ struct CallObjCArcUse final : EHScopeStack::Cleanup {
 
 Value *CodeGenFunction::EmitCheckedArgForBuiltin(const Expr *E,
```
Comment (vitalybuka) [inline clang/lib/CodeGen/CGBuiltin.cpp:1997]: TO my taste, I'd rather forked EmitCheckedArgForBuiltin into a version for Assume
very little code is shared with Zero

File: compiler-rt/lib/ubsan/ubsan_handlers.cpp [hunk 2]
```diff
@@ -633,9 +633,11 @@ static void handleInvalidBuiltin(InvalidBuiltinData *Data, ReportOptions Opts) {
...
   ScopedReport R(Opts, Loc, ET);
 
-  Diag(Loc, DL_Error, ET,
-       "passing zero to %0, which is not a valid argument")
-    << ((Data->Kind == BCK_CTZPassedZero) ? "ctz()" : "clz()");
+  if (Data->Kind == BCK_AssumePassedFalse)
+    Diag(Loc, DL_Error, ET, "assumption is violated during execution");
+  else
+    Diag(Loc, DL_Error, ET, "passing zero to %0, which is not a valid argument")
+        << ((Data->Kind == BCK_CTZPassedZero) ? "ctz()" : "clz()");
```
Comment (vitalybuka) [inline compiler-rt/lib/ubsan/ubsan_handlers.cpp:640]: I suggest introduce `__ubsan::__ubsan_handle_invalid_assume` and we don't need enum here and in Clang.

Also  `__ubsan::__ubsan_handle_invalid_abort` is likely prints garbage.

Would you like to upload a separate patch to fix abort?

**Comment by vitalybuka** [comment]
Thanks!
LGTM in general.

File: compiler-rt/lib/ubsan/ubsan_handlers.cpp [hunk 3]
```diff
@@ -633,9 +633,11 @@ static void handleInvalidBuiltin(InvalidBuiltinData *Data, ReportOptions Opts) {
 
   ScopedReport R(Opts, Loc, ET);
 
-  Diag(Loc, DL_Error, ET,
-       "passing zero to %0, which is not a valid argument")
-    << ((Data->Kind == BCK_CTZPassedZero) ? "ctz()" : "clz()");
+  if (Data->Kind == BCK_AssumePassedFalse)
+    Diag(Loc, DL_Error, ET, "assumption is violated during execution");
```
Comment (vitalybuka) [inline compiler-rt/lib/ubsan/ubsan_handlers.cpp:637]: Maybe keep generic `passing zero to __builtin_assume()` ?

Comment (vitalybuka) [reply compiler-rt/lib/ubsan/ubsan_handlers.cpp:637]: probably fine as is

File: clang/lib/CodeGen/CGBuiltin.cpp (hunk 1 above)
Comment (vitalybuka) [reply clang/lib/CodeGen/CGBuiltin.cpp]: Done

**Comment by vitalybuka** [comment]
I wonder how much code will be broken.
We have internally `fsanitize=builtin` enabled, as is patch will break and force us to disable the sanitizers.
I will try to run internal tests, and clang itself with a new check.

If it breaks to much, we will need to create a new sanitizers, e.g `-fsanitize=assume`.

**Comment by vitalybuka** [comment]
clang with UBSAN is fine, i am running other tests.

**Comment by vitalybuka** [comment]
Other tests are also fine.

**Review by vitalybuka — APPROVED** [review]

**Comment by chapuni** [comment]
This has triggered a failure in our private builder (aarch64-ubuntu20.04)

```
******************** TEST 'libFuzzer-aarch64-default-Linux :: strncmp.test' FAILED ********************
Exit Code: 1
Command Output (stderr):
--
RUN: at line 2: /home/bb/clang-aarch64/build/2/./bin/clang    -Wthread-safety -Wthread-safety-reference -Wthread-safety-beta   --driver-mode=g++ -O2 -gline-tables-only -fsanitize=address,fuzzer -I/home/bb/clang-aarch64/llvm-project/compiler-rt/lib/fuzzer  -Wthread-safety -Wthread-safety-reference -Wthread-safety-beta  /home/bb/clang-aarch64/llvm-project/compiler-rt/test/fuzzer/StrncmpTest.cpp -o /home/bb/clang-aarch64/build/2/projects/compiler-rt/test/fuzzer/AARCH64DefaultLinuxConfig/Output/strncmp.test.tmp-StrncmpTest
+ /home/bb/clang-aarch64/build/2/./bin/clang -Wthread-safety -Wthread-safety-reference -Wthread-safety-beta --driver-mode=g++ -O2 -gline-tables-only -fsanitize=address,fuzzer -I/home/bb/clang-aarch64/llvm-project/compiler-rt/lib/fuzzer -Wthread-safety -Wthread-safety-reference -Wthread-safety-beta /home/bb/clang-aarch64/llvm-project/compiler-rt/test/fuzzer/StrncmpTest.cpp -o /home/bb/clang-aarch64/build/2/projects/compiler-rt/test/fuzzer/AARCH64DefaultLinuxConfig/Output/strncmp.test.tmp-StrncmpTest
RUN: at line 3: not  /home/bb/clang-aarch64/build/2/projects/compiler-rt/test/fuzzer/AARCH64DefaultLinuxConfig/Output/strncmp.test.tmp-StrncmpTest              -seed=2 -runs=10000000   2>&1 | FileCheck /home/bb/clang-aarch64/llvm-project/compiler-rt/test/fuzzer/strncmp.test
+ not /home/bb/clang-aarch64/build/2/projects/compiler-rt/test/fuzzer/AARCH64DefaultLinuxConfig/Output/strncmp.test.tmp-StrncmpTest -seed=2 -runs=10000000
+ FileCheck /home/bb/clang-aarch64/llvm-project/compiler-rt/test/fuzzer/strncmp.test
/home/bb/clang-aarch64/llvm-project/compiler-rt/test/fuzzer/strncmp.test:4:8: error: CHECK: expected string not found in input
CHECK: BINGO
       ^
<stdin>:1:1: note: scanning from here
INFO: Running with entropic power schedule (0xFF, 100).
^
Input file: <stdin>
Check file: /home/bb/clang-aarch64/llvm-project/compiler-rt/test/fuzzer/strncmp.test
-dump-input=help explains the following input dump.
Input was:
<<<<<<
         1: INFO: Running with entropic power schedule (0xFF, 100). 
check:4     X~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ error: no match found
         2: INFO: Seed: 2 
check:4     ~~~~~~~~~~~~~~
         3: INFO: Loaded 1 modules (12 inline 8-bit counters): 12 [0xaaaaccfe3f00, 0xaaaaccfe3f0c),  
check:4     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
         4: INFO: Loaded 1 PC tables (12 PCs): 12 [0xaaaaccfe3f10,0xaaaaccfe3fd0),  
check:4     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
         5: INFO: -max_len is not provided; libFuzzer will not generate inputs larger than 4096 bytes 
check:4     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
         6: INFO: A corpus is not provided, starting from an empty corpus 
check:4     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
         .
         .
         .
>>>>>>
--
********************
```

**Comment by dtcxzyw** [comment]
> -fsanitize=address,fuzzer

I think it is not related to this patch. It only works with `-fsanitize=builtin` or `-fsanitize=undefined`.

**Pull request merged**